[SQL]  BREAKING CHANGE: Update Microsoft.Sql to track2 SDK

ericshape opened the pull request:
**Description**<!--Mandatory-->
Update Microsoft.Sql to track2 Python SDK. 

**Testing Guide**
1. Passed **61/61** end-to-end tests
2. Recorded **60/61** end-to-end tests
3. Skipped **SqlManagedDatabaseLogReplayScenarionTest** in the pipeline and mark it as live_only due there is cred scan issue in the recording file 

**History Notes**
<!--If your PR is not customer-facing, use {Component Name} in the PR title. Otherwise, use [Component Name] to allow our pipeline to add the title as a history note. If you need multiple history notes or would like to overwrite the note from the PR title, please fill in the following templates.-->

[SQL] BREAKING CHANGES: 
1. `sql server ad-admin update` command: add `--administrator-name {administrator_name}` as required parameter
2. `sql server ad-admin delete` command: add `--administrator-name {administrator_name}` as required parameter

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

yonzhan commented:
SQL

yonzhan commented:
Please resolve the conflicting files first.

ericshape commented on src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py:
```diff
@@ -1477,7 +1464,7 @@ def test_sql_db_security_mgmt(self, resource_group, resource_group_2,
         email_account_admins = True
 
-        self.cmd('sql db threat-policy update -g {} -s {} -n {}'
+        self.cmd('sql db threat-policy update -g {} -s {} -n {} --security-alert-policy-name default'
```
@Juliehzl 
I have a question about the newly added `security-alert-policy-name` as required parameter in track2 Python SDK: https://github.com/Azure/azure-sdk-for-python/blob/ede50054a5b78e1a69785221d7da19a0bb817278/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/operations/_database_security_alert_policies_operations.py#L121

this required parameter `security-alert-policy-name` has only one default value "default": https://github.com/Azure/azure-sdk-for-python/blob/ede50054a5b78e1a69785221d7da19a0bb817278/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/models/_sql_management_client_enums.py#L868 . Thus, in track1 python SDK, it was assigned internally: https://github.com/Azure/azure-sdk-for-python/blob/9d8c393590ee726bc0b0d02e8acafa523a79b694/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/operations/_database_security_alert_policies_operations.py#L39

In the commands.py of this PR:
```
        g.generic_update_command('update',
                                 setter_name='create_or_update',
                                 custom_func_name='db_threat_detection_policy_update',
                                 )
```
it seems `generic_update_command` does not support assigning `security-alert-policy-name` without passing in it explicitly as `--security-alert-policy-name default` in the command line. But adding one more `--security-alert-policy-name default` as a mandatory parameter will be the breaking change of the old `sql db threat-policy update` command.

Do you know how can I assign the  `security-alert-policy-name` parameter in `generic_update_command`? 
Thanks!

Juliehzl replied:
Then you need to rewrite create_or_update() and pass security-alert-policy-name in your create_or_update.

ericshape replied:
do you have an example? 
do you mean I can override `setter_name`? 
I assume setter_name is the call path of Python SDK interface and custom_func_name is the method to assign for `parameters` signature in `create_or_update`.

ericshape replied:
cc: @ayeletshpigelman 'cause I might make the breaking change of `sql db threat-policy update` CLI command

evelyn-ys replied:
@ericshape You can simply assign the security-alert-policy-name parameter in the `custom_func` which is `db_threat_detection_policy_update`

PS. You can find the intro for `custom_func_name` of `generic_update_command` [here](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#generic-update-commands)

ericshape replied:
do you know how to update two fields: `security-alert-policy-name` and `parameters` ? 

```
    def generic_update_command(self, name, getter_name='get', getter_type=None,
                               setter_name='create_or_update', setter_type=None, setter_arg_name='parameters',
                               child_collection_prop_name=None, child_collection_key='name', child_arg_name='item_name',
                               custom_func_name=None, custom_func_type=None, **kwargs):
                               
```
                        
via the generic_update_command  impl, it seems the custom_func_name only change `setter_arg_name='parameters'` in default one value ONLY.

```
        if self.setter_arg_name == 'parameters':
            result = cached_put(self.cmd, setter, **setterargs)
        else:
            result = cached_put(self.cmd, setter, setterargs[self.setter_arg_name],
                                setter_arg_name=self.setter_arg_name, **setterargs)             
                               
```

jsntcy replied:
@ericshape, here is an example in `Network` https://github.com/Azure/azure-cli/blob/afec34798bf79439fd49b75a6d82d0ad714a1f05/src/azure-cli/azure/cli/command_modules/network/commands.py#L1418, where you define your own custom function as setter instead of call `begin_create_or_update` function in SDK directly. In your custom function, you call `begin_create_or_update` function in SDK where you can pass both `security-alert-policy-name` (with value `default`) and `parameters`.

![image](https://user-images.githubusercontent.com/16348853/126584163-e331c41e-4870-41ea-9675-b598d403de6c.png)

ericshape replied:
address the comment

ericshape commented:
Hi @Juliehzl 
do you know how to address this issue?
one test was passed and recorded in my local machine; but in the pipeline, it is failed. 

I found there are assert errors, i.e. the resource_id in the record file does not match.

FAILED src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py::SqlDbSensitivityClassificationsScenarioTest::test_sql_db_sensitivity_classifications

`E           /providers/Microsoft.Management/managementGroups/00000000-0000-0000-0000-000000000000/providers/Microsoft.Security/informationProtectionPolicies/effective != /providers/Microsoft.Management/managementGroups/72f988bf-86f1-41af-91ab-2d7cd011db47/providers/Microsoft.Security/informationProtectionPolicies/effective`

https://dev.azure.com/azure-sdk/public/_build/results?buildId=990377&view=logs&j=15e5eb7c-cf66-5b91-00a2-db9b11b5e0f9&t=845fa2ee-a602-59f5-4e4c-07ee6d0283b9&l=3011

Thanks!

ericshape commented on src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py:
```diff
@@ -4182,108 +4183,54 @@ def test_sql_managed_instance_mgmt(self):
-
-        self.cmd('sql mi delete -g {} -n {} --yes'
-                 .format(resource_group_1, managed_instance_name), checks=NoneCheck())
+                     JMESPathCheck('primaryUserAssignedIdentityId', self.test_umi),
```
for this test, I also met the same situation as I commented in: https://github.com/Azure/azure-cli/pull/18637#issuecomment-876825197

I passed this test locally. but it failed in pipeline due to empty guide:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=999285&view=logs&j=15e5eb7c-cf66-5b91-00a2-db9b11b5e0f9&t=845fa2ee-a602-59f5-4e4c-07ee6d0283b9&l=5679

E               azure.cli.testsdk.exceptions.JMESPathCheckAssertionError: Query 'primaryUserAssignedIdentityId' doesn't yield expected value '/subscriptions/e64f3e8e-ab91-4a65-8cdd-5cd2f47d00b4/resourcegroups/viparek/providers/Microsoft.ManagedIdentity/userAssignedIdentities/testumi', instead the actual value is '/subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/viparek/providers/Microsoft.ManagedIdentity/userAssignedIdentities/testumi'. Data: 


do we have any way to resolve this issue?

Juliehzl replied:
There are two ways:
1. do not check full resource id for user assigned identity. Name is enough.
2. Replace subscription information with 00000000-0000-0000-0000-000000000000 in playback mode.

ericshape replied:
addressed with if condition of is_live/is_recording

ericshape commented:
@yonzhan @Juliehzl 
can you help to answer the questions above? Thanks!

evelyn-ys commented on src/azure-cli/azure/cli/command_modules/sql/_params.py:
```diff
@@ -940,6 +941,13 @@ def _configure_security_policy_storage_params(arg_ctx):
     with self.argument_context('sql db threat-policy update') as c:
         _configure_security_policy_storage_params(c)
 
+        c.argument('security_alert_policy_name',
+                   arg_group=policy_arg_group,
+                   options_list=['--security-alert-policy-name', '-p'],
+                   required=True,
+                   help='The name of the security alert policy. The value is Default.',
+                   arg_type=get_enum_type(SecurityAlertPolicyName))
```
Does this parameter have only one allowed value `default`? Can it be added inside cli custom function so that cli customer won't face any breaking change?

ericshape replied:
yes. the `default` only `security_alert_policy_name` is the new added parameter in track2 SDK. 
previously, in track1 SDK, if the enum has only one value, it was assigned as `default` in SDK itself. do you think it is a bug of Python SDK? 

In track2 SDK, the update method interface has changed to https://github.com/Azure/azure-sdk-for-python/blob/ede50054a5b78e1a69785221d7da19a0bb817278/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/operations/_database_security_alert_policies_operations.py#L121

with the `custom_update_command` wrap, the customized func can only change the `parameter`, not `security_alert_policy_name`. Thus, I have to make breaking change for
> sql db threat-policy update command: add --security-alert-policy-name default as required parameter

evelyn-ys replied:
Replied https://github.com/Azure/azure-cli/pull/18637#discussion_r674430846

ericshape replied:
addressed the comment

jsntcy commented on src/azure-cli/azure/cli/command_modules/sql/commands.py:
```diff
@@ -315,7 +324,7 @@ def load_command_table(self, _):
 
         g.custom_command('set', 'update_long_term_retention')
-        g.show_command('show', 'get')
+        g.custom_show_command('show', 'update_long_term_retention_get')
```
Better name `get_long_term_retention`

ericshape replied:
addressed

jsntcy commented on src/azure-cli/azure/cli/command_modules/sql/commands.py:
```diff
@@ -772,8 +796,8 @@ def load_command_table(self, _):
 
-        g.custom_command('set', 'update_long_term_retention_mi', is_preview=True)
-        g.show_command('show', 'get', is_preview=True)
+        g.custom_show_command('set', 'update_long_term_retention_mi', is_preview=True)
```
Use `g.custom_command(`set`, xxx)` for `set` command as `custom_show_command` is only for `show` command`.

ericshape replied:
addressed

jsntcy commented on src/azure-cli/azure/cli/command_modules/sql/commands.py:
```diff
@@ -610,8 +630,8 @@ def load_command_table(self, _):
-        c.show_command('show', 'get')
-        c.generic_update_command('update')
+        c.custom_show_command('show', 'conn_policy_show')
+        c.custom_command('update', 'conn_policy_update')
```
Minor issue: function naming is not consistent in SQL, some are named as `noun_verb`, the others are named as `verb_noun`. Usually we name it as `verb_noun`, for example `update_conn_policy`.

ericshape replied:
addressed

jsntcy commented:
@ericshape Overall LGTM except some minor issues which I already added comments.

ericshape commented:
> @ericshape Overall LGTM except some minor issues which I already added comments.

@jsntcy I have addressed your comments. 
now the only discussion left is https://github.com/Azure/azure-cli/pull/18637#discussion_r673780968

ericshape commented:
@evelyn-ys  @Juliehzl  can you help to review and merge this PR? Thanks!

evelyn-ys reviewed (approved):
LGTM except for the skipped AAD tests
